assets/js/data.coffee

```diff
@@ -249,11 +263,65 @@ class Data
   toggleCollapsed: (row) ->
     @store.setCollapsed row.id, (not @collapsed row)
 
+  countInstances: (id) ->
+    # Precondition: No circular references in ancestry
+    errors.assert id?, "Empty id passed to countInstances"
+    if id == constants.root_id
+      return 1
+    parentCount = 0
+    for parent_id in (@store.getParents id)
+      parentCount += @countInstances parent_id # Always exactly once under every parent
+    return parentCount
```

Comment:
1. exactlyOneInstance will return the incorrect result if it literally only checks the number of parents, although it could be special-cased, yes.
2. this isn't really a very expensive operation, could you be more specific as to how you think it will take a long time? if we added memoization to the parent count, it would be O(ancestors) which is optimal for exactlyOneInstance. If we don't it double-counts some diamond-shaped ancestry but it's still quite fast.